### ISSUE TITLE
Webhook-service upgrade service driver

### ISSUE THREAD

**Issue by mrajashree** [rancher/rancher#7735]
Add driver for upgrading services based on Dockerhub webhooks</br>
**Creating and using the webhook**
1. This webhook will accept three fields: `image`, `serviceSelector` and `tag`.
2. `serviceSelector` is the service label. Using this we can have single webhook for multiple services. So all services with label `foo=bar` can get upgraded by the same webhook.
3. The url (`/v1-webhooks/endpoint?key=value`) from created webhook should be used as the Dockerhub webhook.
4. When an image is pushed to Dockerhub, it sends POST request with body containing the pushed image name and tag to the Dockerhub webhook (Url from the created webhook)

**Upgrade scenarios**
Upgrade will take place only if the pushed tag matches the current tag
So if pushed tag == `14.04` and current is `14.04`, then upgrade
If pushed tag == `latest` and current `latest`, then upgrade

**Comment by mrajashree** [comment]
- [ ] Set label for pulling image before create to "always" during upgrade

- [ ] Use uuid & projectId encoded in the webhook url in place of jwt

**Comment by sangeethah** [comment]
@mrajashree:

1. What values will be set for ```Batch Size``` , ```Batch Interval``` and ```Start Before/after Stopping``` ?
2. Would it be better to allow users to set the above values ? ``Start Before/after Stopping``` would be of importance when there are exposed ports and we do not have enough number (twice the scale) of hosts in the setup.
3. Since there can be multiple services selected for upgrade , Would the upgrades of the services be done in parallel or in sequence ?
4. How is the status of each service upgrade communicated to the user ?
5.  In case of services with sidekicks , How is the decision on what set of services ( primary vs sidekicks) to be picked for upgrade made ?

**Comment by mrajashree** [comment]
1. Batch size is 1 and rest are default values currently.
2. Not sure
3. Upgrades are sequential right now
4. Currently webhook-service is only sending error messages, and 200 for successful upgrades. Do we also want to communicate successful upgrades @cjellick ?
5. Not sure

**Comment by sangeethah** [comment]
@mrajashree , users would have specific requirement  for  Batch Size , Batch Interval and Start Before/after Stopping for managing the service downtime. It would be good to allow users to specify these parameters.

**Comment by mrajashree** [comment]
@sangeethah yeah, since they can specify these while manually doing an upgrade, makes sense to have these options for them during webhook creation. But we can select multiple services while creating a webhook, so these parameters would be used for upgrade of all these services. That should be okay right?

**Comment by cjellick** [comment]
My input:
> If service not in active state, can we upgrade?
If the API allows it, I'd allow it in the webhook

>  If upgrade fails, do we rollback?
I would not rollback. I'd leave the service in the state it is in.

> 1. What values will be set for ```Batch Size``` , ```Batch Interval``` and ```Start Before/after Stopping``` ?
I like this. Let's add it. @mrajashree regarding your Q:
> ...so these parameters would be used for upgrade of all these services. That should be okay right?
Yes, that should be ok.

> 2. Would it be better to allow users to set the above values ? ```Start Before/after Stopping``` would be of importance when there are exposed ports and we do not have enough number (twice the scale) of hosts in the setup.
>> 2. Not sure
Yeah, let's let them set them.

> 3. Since there can be multiple services selected for upgrade , Would the upgrades of the services be done in parallel or in sequence ?
>> 3. Upgrades are sequential right now
This is fine for now. We can move to parallel in the future if need be.

> 4. How is the status of each service upgrade communicated to the user ?
>> 4. Currently webhook-service is only sending error messages, and 200 for successful upgrades. Do we also want to communicate successful upgrades @cjellick ?
Yeah, the point of the webhook is that the thing calling it is a piece of software that effectively fires-and-forgets.

> 5.  In case of services with sidekicks , How is the decision on what set of services ( primary vs sidekicks) to be picked for upgrade made ?
>> 5. Not sure
I don't quite understand this question. All services that match the label will be upgraded. If the sidekick doesn't have the matching label, it won't be upgraded.

> 6. This feature should work with private registry as well?  
>>  6. Doesn't work for private repo
To explain: the webhook relies on the data payload sent by Docker Hub to figure out what image tag was pushed. If we get demand for "support private repo X" in the future, we can add logic to support that.

**Comment by sangeethah** [comment]
@cjellick @mrajashree Thank you for the input.

Few follow up questions:

1. My intent with asking about private repo was actually private image. We should be able to auto upgrade using private images in docker hub , right  (as long as we have credentials stored in our registries) ?

2. Since upgrades are being done sequence , do we have any times outs in place to  consider an upgrade process that is stuck in "upgrading" state to be skipped and move on to the next service that needs to be upgraded?

**Comment by mrajashree** [comment]
@sangeethah @cjellick 
1. I hadn't added registry with my docker hub credentials yesterday while trying upgrade with a private image, my bad. I tried that and it works.

2. There is a wait logic in the code, but it returns error for one service and webhook-service returns an error then. Instead I can change it to not returning an error till upgrade/finishUpgrade is called for all services? @cjellick

**Comment by sangeethah** [comment]
@cjellick @mrajashree , since we are setting label for pulling image before create to "always" during upgrade, we expect as many image pulls to happen as there are number of containers(part of service that are identified for upgrade) . Many of these pulls may be pulling the same tag of the image if they happen to end on the same host as other containers that were upgraded earlier during this process.

Is pulling the image once on all hosts for every WR an option to consider in the place of having "always" pull for the upgrade request? yes , we may end up pulling images in hosts where no containers would get deployed. This may be something to consider in cases where there are lot of containers qualifying for upgrades .

**Comment by mrajashree** [comment]
@sangeethah I might be wrong but I don't think I can pull image on host using cattle api calls. And if a user wanted to upgrade all these services in response to Docker Hub webhook in absence of webhook-service, even then he'd pull the image for every service as it is to get the updated image. cc @cjellick

**Comment by cjellick** [comment]
Yep @sangeethah I agree that better optimized image pulling is a good idea and can be done in the future.

**Comment by soumyalj** [comment]
Tested with master. Feature to upgrade services through webhooks is functional.

**Comment by Negashev** [comment]
v1.5.5 Batch Interval `ignored`

**Comment by cjellick** [comment]
@Negashev I'm going to open a new issue for this
